pkg/executor/aggregate/agg_hash_partial_worker.go

```diff
@@ -67,43 +87,140 @@ func (w *HashAggPartialWorker) getChildInput() bool {
 	return true
 }
 
+func (w *HashAggPartialWorker) fetchChunkAndProcess(ctx sessionctx.Context, hasError *bool, needShuffle *bool, enableIntest bool) bool {
+	w.spillHelper.syncLock.RLock()
+	defer w.spillHelper.syncLock.RUnlock()
+
+	if w.spillHelper.isInSpilling() {
+		// Repeatedly release lock so that it could stop processing data when
+		// spill is in execution and spill action could have more chance to get write lock.
+		return true
+	}
+
+	waitStart := time.Now()
+	ok := w.getChildInput()
+	if w.stats != nil {
+		w.stats.WaitTime += int64(time.Since(waitStart))
+	}
+
+	if !ok {
+		return false
+	}
+
+	execStart := time.Now()
+	if err := w.updatePartialResult(ctx, w.chk, len(w.partialResultsMap)); err != nil {
+		*hasError = true
+		w.processError(err)
+		return false
+	}
+	if w.stats != nil {
+		w.stats.ExecTime += int64(time.Since(execStart))
+		w.stats.TaskNum++
+	}
+
+	// The intermData can be promised to be not empty if reaching here,
+	// so we set needShuffle to be true.
+	*needShuffle = true
+
+	if intest.InTest && enableIntest {
+		num := rand.Intn(10000)
+		if num < 7 {
+			panic("Intest panic: partial worker is panicked when running")
+		} else if num < 14 {
+			time.Sleep(1 * time.Millisecond)
+		} else if num < 21 {
+			*hasError = true
+			w.processError(errors.Errorf("Random fail is triggered in partial worker"))
+			return false
+		} else if num < 28 {
+			w.memTracker.Consume(500000)
+		}
+	}
+
+	if w.spillHelper.checkError() {
+		*hasError = true
+		return false
+	}
+	return true
+}
+
+func (w *HashAggPartialWorker) handleSpillBeforeExit() {
+	w.spillHelper.syncLock.RLock()
+	defer w.spillHelper.syncLock.RUnlock()
+
+	if w.spillHelper.checkError() {
+		return
+	}
+
+	hasError := false
+	if len(w.groupKey) > 0 {
+		if err := w.spillDataToDisk(); err != nil {
+			w.processError(err)
+			hasError = true
+		}
+	}
+
+	if !hasError {
+		w.spillHelper.addListInDisks(w.spilledChunksIO)
+	}
+}
+
+func (w *HashAggPartialWorker) sendDataToFinalWorkersBeforeExit(needShuffle bool, finalConcurrency int, hasError bool) {
+	w.runningWorkerWaiter.Done()
+	w.runningWorkerWaiter.Wait()
```

Comment:
Will it be some potential hang issue if some partialWorker fail to start but the `runningWorkerWaiter` is already inited to have `partialConcurrency`?

Reply:
> Will it be some potential hang issue if some partialWorker fail to start but the `runningWorkerWaiter` is already inited to have `partialConcurrency`?

Only when some of goroutine fail to be created the hang will happen, so I think it's safety enough.

Reply:
It's a little werid to me that `HashAggPartialWorker` wait all `HashAggPartialWorker` done when it is still in running state.

Reply:
> It's a little werid to me that `HashAggPartialWorker` wait all `HashAggPartialWorker` done when it is still in running state.

Because some partial worker may finish task early without spill triggered, it will send its data to final worker, after a while, the spill may be triggered. So we can know if the spill is triggered only when all partial workers finish their tasks then can we know how to handle the partial results.

Reply:
According to your explaination, I suggest to rename it to something like `paralleSpillWaiter`?

Reply:
> According to your explaination, I suggest to rename it to something like `paralleSpillWaiter`?

okk